README.md
# cloud-starter

本插件是基于springcloud封装的远程服务调用组件,旨在简化springcloud原生的远程服务调用代码复杂度(服务端不用写controller发布服务,客户端不用使用feign或者ribbon注册服务),从而提升开发效率.

本插件只封装简化了springcloud的"服务注册与发现",不影响其他功能的使用,同时也不影响程序员继续使用原生的"服务注册与发现".

本插件提示了跨平台的日志追踪功能.只需要@EnableSeriesLogs简单的一个注解,就可以实现同一次请求的不同平台的日志具有相同的前缀,方便日志追踪.

插件由 客户端client-starter插件和服务端server-starter插件两部分组成.两个组件配合使用.

注意:使用了本插件,无需在进行springcloud相关maven依赖的的的引用和相关注解的开启.           



使用方法:

1.服务端发布服务

	1.1 pom引入maven依赖   
		<dependency>
			<groupId>com.lianshang.cloud</groupId>
			<artifactId>cloud-server-starter</artifactId>
			<version>${cloud-server-starter.version}</version>
		</dependency>

		当前最新版本 0.0.2



	1.2  对service实现类添加注解 @LsCloudService
	例如:
	/**
	 * @author: lianshang
	 **/
	@LsCloudService
	public class PayServiceImpl implements IPayService {

	.....


	}

	此刻,已经完成了服务端service的发布.



2.客户端使用远程服务

	2.1 pom 引入maven依赖
	<dependency>
		<groupId>com.lianshang.cloud</groupId>
		<artifactId>cloud-client-starter</artifactId>
		<version>${cloud-client-starter.version}</version>
	</dependency>

	当前最新版本 0.0.2


	2.2 客户端使用远程service服务,直接用 @LsCloudAutowired("服务名称")

	@RestController
	@RequestMapping("order")
	public class OrderController {

  	@LsCloudAutowired("CLOTH2-BIZ-SERVICE") #####CLOTH2-BIZ-SERVICE 为eureka注册中心中服务(应用)名称
  	private OrderService orderService;

  	.....

  	}



3.提供了跨平台的日志串联(非必须,只有需要开启跨平台日志,才进行配置)

  基于MDC提供了跨平台连续日志的功能,便于日志的追踪.同一个平台中,同一次请求的产生的日志都有相同的前缀;跨平台的远程服务调用,如果客户端和服务端都开启了跨平台连续日志功能.不同平台的日志也具有相同的前缀,从而便于不同平台的同一次请求的日志追踪.

  3.1 服务端或者客户端使用 @EnableSeriesLogs注解可以开启当前平台的串联日志

  例如:

    .....
  	@EnableSeriesLogs
	public class SharpWebConfig implements WebMvcConfigurer {
      .........

      }



  3.2 logback.xml(或者log4j)的日志配置中添加 %X{lsReq}
  
  例如:
   <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="ch.qos.logback.classic.encoder.PatternLayoutEncoder">
            <pattern>%X{lsReq} %d{yyyy-MM-dd HH:mm:ss.SSS} [%thread] %-5level %logger -%line - %msg%n</pattern>
        </encoder>
    </appender>
    
    
  打印日志举例:

	  客户端日志:
	 【a660f9b2-a669-4b56-b585-1b8369d4efcc_-55】 2018-12-17 15:00:36.809 [http-nio-8080-exec-1] INFO  com.lianshang.common.controller.aspect.HandleControllerAspect -59 - 请求信息===> className:BasicErrorController,methodName:errorHtml,args:[org.apache.catalina.core.ApplicationHttpRequest@3529f16a, org.apache.catalina.connector.ResponseFacade@2152dbd3]
	 【a660f9b2-a669-4b56-b585-1b8369d4efcc_-55】 2018-12-17 15:00:36.943 [http-nio-8080-exec-1] INFO  com.lianshang.common.controller.aspect.	  HandleControllerAspect -97 - 响应信息===> className:BasicErrorController,methodName:errorHtml,入参:[org.apache.catalina.core.ApplicationHttpRequest@3529f16a, org.apache.catalina.connector.ResponseFacade@2152dbd3], 返回结果:null, 耗时:136


    服务端日志:
   【a660f9b2-a669-4b56-b585-1b8369d4efcc_-55】 2018-12-17 15:00:36.809 [http-nio-8080-exec-1] INFO  com.lianshang.common.controller.aspect.HandleControllerAspect -59 - 请求信息===> className:BasicErrorController,methodName:errorHtml,args:[org.apache.catalina.core.ApplicationHttpRequest@3529f16a, org.apache.catalina.connector.ResponseFacade@2152dbd3]
   【a660f9b2-a669-4b56-b585-1b8369d4efcc_-55】 2018-12-17 15:00:36.943 [http-nio-8080-exec-1] INFO  com.lianshang.common.controller.aspect.HandleControllerAspect -97 - 响应信息===> className:BasicErrorController,methodName:errorHtml,入参:[org.apache.catalina.core.ApplicationHttpRequest@3529f16a, org.apache.catalina.connector.ResponseFacade@2152dbd3], 返回结果:null, 耗时:136
   日志虽然在不同平台,但是有着相同的前缀  【a660f9b2-a669-4b56-b585-1b8369d4efcc_-55】



CHANGELOG

0.0.1  第一版发布

0.0.2  1.修复接口(父类)泛型类型 反序列时真实类型丢失导致反序列失败的问题.

        例如:
        public class ServiceImpl<M extends LsBaseMapper<T>, T, DTO> implements IService<DTO> {
            ...
                public DTO getById(Serializable id) {
                    ...
                }
            ...
        }
        公共的getList,getById方法的反序列化

       2. 解决了入参中存在null时,获取真实参数类型失败,导致的空指针问题




lsCloudStarter-server/src/main/java/com/lianshang/cloud/server/utils/CRC8Util.java
package com.lianshang.cloud.server.utils;
/**
 * CRC工具类
 * @author 孙龙云
 */
public class CRC8Util {
	static byte[] crc8_tab = {(byte) 0, (byte) 94, (byte) 188, (byte) 226, (byte) 97, (byte) 63, (byte) 221, (byte) 131, (byte) 194, (byte) 156, (byte) 126, (byte) 32, (byte) 163, (byte) 253, (byte) 31, (byte) 65, (byte) 157, (byte) 195, (byte) 33, (byte) 127, (byte) 252, (byte) 162, (byte) 64, (byte) 30, (byte) 95, (byte) 1, (byte) 227, (byte) 189, (byte) 62, (byte) 96, (byte) 130, (byte) 220, (byte) 35, (byte) 125, (byte) 159, (byte) 193, (byte) 66, (byte) 28, (byte) 254, (byte) 160, (byte) 225, (byte) 191, (byte) 93, (byte) 3, (byte) 128, (byte) 222, (byte) 60, (byte) 98, (byte) 190, (byte) 224, (byte) 2, (byte) 92, (byte) 223, (byte) 129, (byte) 99, (byte) 61, (byte) 124, (byte) 34, (byte) 192, (byte) 158, (byte) 29, (byte) 67, (byte) 161, (byte) 255, (byte) 70, (byte) 24,
            (byte) 250, (byte) 164, (byte) 39, (byte) 121, (byte) 155, (byte) 197, (byte) 132, (byte) 218, (byte) 56, (byte) 102, (byte) 229, (byte) 187, (byte) 89, (byte) 7, (byte) 219, (byte) 133, (byte) 103, (byte) 57, (byte) 186, (byte) 228, (byte) 6, (byte) 88, (byte) 25, (byte) 71, (byte) 165, (byte) 251, (byte) 120, (byte) 38, (byte) 196, (byte) 154, (byte) 101, (byte) 59, (byte) 217, (byte) 135, (byte) 4, (byte) 90, (byte) 184, (byte) 230, (byte) 167, (byte) 249, (byte) 27, (byte) 69, (byte) 198, (byte) 152, (byte) 122, (byte) 36, (byte) 248, (byte) 166, (byte) 68, (byte) 26, (byte) 153, (byte) 199, (byte) 37, (byte) 123, (byte) 58, (byte) 100, (byte) 134, (byte) 216, (byte) 91, (byte) 5, (byte) 231, (byte) 185, (byte) 140, (byte) 210, (byte) 48, (byte) 110, (byte) 237,
            (byte) 179, (byte) 81, (byte) 15, (byte) 78, (byte) 16, (byte) 242, (byte) 172, (byte) 47, (byte) 113, (byte) 147, (byte) 205, (byte) 17, (byte) 79, (byte) 173, (byte) 243, (byte) 112, (byte) 46, (byte) 204, (byte) 146, (byte) 211, (byte) 141, (byte) 111, (byte) 49, (byte) 178, (byte) 236, (byte) 14, (byte) 80, (byte) 175, (byte) 241, (byte) 19, (byte) 77, (byte) 206, (byte) 144, (byte) 114, (byte) 44, (byte) 109, (byte) 51, (byte) 209, (byte) 143, (byte) 12, (byte) 82, (byte) 176, (byte) 238, (byte) 50, (byte) 108, (byte) 142, (byte) 208, (byte) 83, (byte) 13, (byte) 239, (byte) 177, (byte) 240, (byte) 174, (byte) 76, (byte) 18, (byte) 145, (byte) 207, (byte) 45, (byte) 115, (byte) 202, (byte) 148, (byte) 118, (byte) 40, (byte) 171, (byte) 245, (byte) 23, (byte) 73, (byte) 8,
            (byte) 86, (byte) 180, (byte) 234, (byte) 105, (byte) 55, (byte) 213, (byte) 139, (byte) 87, (byte) 9, (byte) 235, (byte) 181, (byte) 54, (byte) 104, (byte) 138, (byte) 212, (byte) 149, (byte) 203, (byte) 41, (byte) 119, (byte) 244, (byte) 170, (byte) 72, (byte) 22, (byte) 233, (byte) 183, (byte) 85, (byte) 11, (byte) 136, (byte) 214, (byte) 52, (byte) 106, (byte) 43, (byte) 117, (byte) 151, (byte) 201, (byte) 74, (byte) 20, (byte) 246, (byte) 168, (byte) 116, (byte) 42, (byte) 200, (byte) 150, (byte) 21, (byte) 75, (byte) 169, (byte) 247, (byte) 182, (byte) 232, (byte) 10, (byte) 84, (byte) 215, (byte) 137, (byte) 107, 53};

    /**
     * 计算数组的CRC8校验值
     *
     * @param data 需要计算的数组
     * @return CRC8校验值
     */
    public static byte calcCrc8(byte[] data) {
        return calcCrc8(data, 0, data.length - 1, (byte) 0);
    }

    /**
     * 计算CRC8校验值
     *
     * @param data   数据
     * @param offset 起始位置
     * @param len    长度
     * @return 校验值
     */
    public static byte calcCrc8(byte[] data, int offset, int len) {
        return calcCrc8(data, offset, len, (byte) 0);
    }

    /**
     * 计算CRC8校验值
     *
     * @param data   数据
     * @param offset 起始位置
     * @param len    长度
     * @param preval 之前的校验值
     * @return 校验值
     */
    public static byte calcCrc8(byte[] data, int offset, int len, byte preval) {
        byte ret = preval;
        for (int i = offset; i < (offset + len); ++i) {
            ret = crc8_tab[(0x00ff & (ret ^ data[i]))];
        }
        return ret;
    }
}
